### ISSUE TITLE
Unable to resolve type: Mono.Runtime. at NLog.Internal.PlatformDetector.get_IsMono()

### ISSUE THREAD

**Issue by sylvainclb** [NLog/NLog#4918]
Hello.

I hit an issue similar to #2416, where NLog try to verify if  I am on a Mono Platform
Context: working on a .net 3,5 workflow.

**The current NLog version:** 
4.7.15

**The error message and stacktrace:**
```
Unable to resolve type: Mono.Runtime.
```
```
   at System.Environment.get_StackTrace()&#xD;&#xA;
   at System.Diagnostics.TraceEventCache.get_Callstack()&#xD;&#xA;
   at System.Diagnostics.XmlWriterTraceListener.WriteFooter(TraceEventCache eventCache)&#xD;&#xA;
   at Microsoft.IdentityManagement.CircularTraceListener.CircularXmlTraceListener.TraceEvent(TraceEventCache eventCache, String source, TraceEventType eventType, Int32 id, String message)&#xD;&#xA;
   at System.Diagnostics.TraceSource.TraceEvent(TraceEventType eventType, Int32 id, String message)&#xD;&#xA;
   at Microsoft.IdentityManagement.Logging.IdentityManagementTraceSource.ReportTrace(TraceEventType traceEventType, Int32 eventId, String message)&#xD;&#xA;
   at Microsoft.ResourceManagement.Utilities.LoggingManager.WriteTraceMessage(TraceEventType eventType, EventIdentifier identifier, String message)&#xD;&#xA;
   at Microsoft.ResourceManagement.Utilities.LoggingManager.ReportWarning(String msg, String source, String warningCode, String stackTrace)&#xD;&#xA;
   at Microsoft.ResourceManagement.Utilities.LoggingManager.ReportWarning(String msg)&#xD;&#xA;
   at Microsoft.ResourceManagement.Utilities.ExceptionHandler.TypeResolveEventHandler(Object sender, ResolveEventArgs args)&#xD;&#xA;
   at System.AppDomain.OnTypeResolveEvent(String typeName)&#xD;&#xA;
   at System.RuntimeTypeHandle._GetTypeByName(String name, Boolean throwOnError, Boolean ignoreCase, Boolean reflectionOnly, StackCrawlMark&amp; stackMark, Boolean loadTypeFromPartialName)&#xD;&#xA;
   at System.RuntimeTypeHandle.GetTypeByName(String name, Boolean throwOnError, Boolean ignoreCase, Boolean reflectionOnly, StackCrawlMark&amp; stackMark)&#xD;&#xA;
   at System.RuntimeType.PrivateGetType(String typeName, Boolean throwOnError, Boolean ignoreCase, Boolean reflectionOnly, StackCrawlMark&amp; stackMark)&#xD;&#xA;
   at System.Type.GetType(String typeName)&#xD;&#xA;
   at NLog.Internal.PlatformDetector.get_IsMono()&#xD;&#xA;
   at NLog.Internal.PlatformDetector.get_SupportsSharableMutex()&#xD;&#xA;
   at NLog.Internal.FileAppenders.BaseMutexFileAppender..ctor(String fileName, ICreateFileParameters createParameters)&#xD;&#xA;
   at NLog.Internal.FileAppenders.RetryingMultiProcessFileAppender.Factory.NLog.Internal.FileAppenders.IFileAppenderFactory.Open(String fileName, ICreateFileParameters parameters)&#xD;&#xA;
   at NLog.Internal.FileAppenders.FileAppenderCache.CreateAppender(String fileName, Int32 freeSpot)&#xD;&#xA;
   at NLog.Internal.FileAppenders.FileAppenderCache.AllocateAppender(String fileName)&#xD;&#xA;
   at NLog.Targets.FileTarget.WriteToFile(String fileName, ArraySegment`1 bytes, Boolean initializedNewFile)&#xD;&#xA;
   at NLog.Targets.FileTarget.ProcessLogEvent(LogEventInfo logEvent, String fileName, ArraySegment`1 bytesToWrite)&#xD;&#xA;
   at NLog.Targets.FileTarget.Write(LogEventInfo logEvent)&#xD;&#xA;
   at NLog.Targets.Target.Write(AsyncLogEventInfo logEvent)&#xD;&#xA;
   at NLog.Targets.Target.WriteAsyncThreadSafe(AsyncLogEventInfo logEvent)&#xD;&#xA;
   at NLog.Targets.Target.WriteAsyncLogEvent(AsyncLogEventInfo logEvent)&#xD;&#xA;
   at NLog.LoggerImpl.WriteToTargetWithFilterChain(Target target, FilterResult result, LogEventInfo logEvent, AsyncContinuation onException)&#xD;&#xA;
   at NLog.LoggerImpl.Write(Type loggerType, TargetWithFilterChain targetsForLevel, LogEventInfo logEvent, LogFactory factory)&#xD;&#xA;
   at NLog.Logger.WriteToTargets(LogLevel level, IFormatProvider formatProvider, String message)&#xD;&#xA;
   at MY.NAMESPACE.CLASS..ctor()&#xD;&#xA;
   at System.RuntimeTypeHandle.CreateInstance(RuntimeType type, Boolean publicOnly, Boolean noCheck, Boolean&amp; canBeCached, RuntimeMethodHandle&amp; ctor, Boolean&amp; bNeedSecurityCheck)&#xD;&#xA;
   at System.RuntimeType.CreateInstanceSlow(Boolean publicOnly, Boolean fillCache)&#xD;&#xA;
   at System.RuntimeType.CreateInstanceImpl(Boolean publicOnly, Boolean skipVisibilityChecks, Boolean fillCache)&#xD;&#xA;
   at System.Activator.CreateInstance(Type type, Boolean nonPublic)&#xD;&#xA;
   at System.Workflow.ComponentModel.Serialization.WorkflowMarkupSerializer.CreateInstance(WorkflowMarkupSerializationManager serializationManager, Type type)&#xD;&#xA;
   at System.Workflow.ComponentModel.Serialization.ActivityMarkupSerializer.CreateInstance(WorkflowMarkupSerializationManager serializationManager, Type type)&#xD;&#xA;
   at System.Workflow.ComponentModel.Serialization.WorkflowMarkupSerializer.CreateInstance(WorkflowMarkupSerializationManager serializationManager, XmlQualifiedName xmlQualifiedName, XmlReader reader)&#xD;&#xA;
   at System.Workflow.ComponentModel.Serialization.WorkflowMarkupSerializer.DeserializeObject(WorkflowMarkupSerializationManager serializationManager, XmlReader reader)&#xD;&#xA;
   at System.Workflow.ComponentModel.Serialization.WorkflowMarkupSerializer.DeserializeContents(WorkflowMarkupSerializationManager serializationManager, Object obj, XmlReader reader)&#xD;&#xA;
   at System.Workflow.ComponentModel.Serialization.WorkflowMarkupSerializer.DeserializeObject(WorkflowMarkupSerializationManager serializationManager, XmlReader reader)&#xD;&#xA;
   at System.Workflow.ComponentModel.Serialization.WorkflowMarkupSerializer.DeserializeXoml(WorkflowMarkupSerializationManager serializationManager, XmlReader xmlReader)&#xD;&#xA;
   at System.Workflow.ComponentModel.Serialization.WorkflowMarkupSerializer.Deserialize(IDesignerSerializationManager serializationManager, XmlReader reader)&#xD;&#xA;
   at System.Workflow.ComponentModel.Serialization.WorkflowMarkupSerializer.Deserialize(XmlReader reader)&#xD;&#xA;
   at Microsoft.ResourceManagement.Workflow.Hosting.HostActivator.DeserializeXomlDefinition(String xomlDefinition)&#xD;&#xA;
   at Microsoft.ResourceManagement.Workflow.Hosting.HostActivator.WorkflowRequiresServiceHost(String xomlDefinition)&#xD;&#xA;
   at Microsoft.ResourceManagement.Workflow.Hosting.HostActivator.ActivateHost(ResourceManagementWorkflowDefinition workflowDefinition, Boolean suspendWorkflowStartupAndTimerOperations)&#xD;&#xA;
   at Microsoft.ResourceManagement.Workflow.Hosting.HostActivator.RetrieveWorkflowDataForHostActivator()&#xD;&#xA;
   at Microsoft.ResourceManagement.Workflow.Hosting.HostActivator.UpdateServiceHosts(Object source, ElapsedEventArgs e)&#xD;&#xA;
   at System.Timers.Timer.MyTimerCallback(Object state)&#xD;&#xA;
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)&#xD;&#xA;
   at System.Threading._TimerCallback.PerformTimerCallback(Object state)
```

**The internal log, Debug level. :**
```
2022-05-12 12:51:39.4813 Debug FileTarget(Name=logfile_mylog): Preparing for new file 'F:\LOGS\MYLOG.log'
2022-05-12 12:51:39.5184 Debug Creating file appender: F:\LOGS\MYLOG.log
2022-05-12 12:52:29.2300 Debug Setting 'NLog.LayoutRenderers.Wrappers.UppercaseLayoutRendererWrapper.uppercase' to 'true'
2022-05-12 12:52:29.2300 Debug Setting 'NLog.LayoutRenderers.DateLayoutRenderer.format' to 'yyyy-MM-dd HH:mm:ss'
2022-05-12 12:52:29.2300 Debug Setting 'NLog.LayoutRenderers.Wrappers.UppercaseLayoutRendererWrapper.uppercase' to 'true'
2022-05-12 12:52:29.2300 Debug Setting 'NLog.LayoutRenderers.Wrappers.PaddingLayoutRendererWrapper.padding' to '-5'
2022-05-12 12:52:29.2300 Debug Registered target logfile_mylog: NLog.Targets.FileTarget
2022-05-12 12:52:29.2300 Info Closing old configuration.
2022-05-12 12:52:29.2300 Debug LogFactory Flush with timeout=15 secs
2022-05-12 12:52:29.2510 Debug Flush completed
2022-05-12 12:52:29.2510 Debug Closing logging configuration...
2022-05-12 12:52:29.2510 Debug Closing target 'File Target[logfile_mylog]'.
2022-05-12 12:52:29.2620 Debug FileAppender Closing Invalidate - F:\LOGS\MYLOG.log
2022-05-12 12:52:29.2620 Debug Closed target 'File Target[logfile_mylog]'.
2022-05-12 12:52:29.2620 Debug Finished closing logging configuration.
2022-05-12 12:52:29.2620 Debug --- NLog configuration dump ---
2022-05-12 12:52:29.2620 Debug Targets:
2022-05-12 12:52:29.2620 Debug File Target[logfile_mylog]
2022-05-12 12:52:29.2620 Debug Rules:
2022-05-12 12:52:29.2620 Debug logNamePattern: (:All) levels: [ Debug Info Warn Error Fatal ] appendTo: [ logfile_mylog ]
2022-05-12 12:52:29.2620 Debug --- End of NLog configuration dump ---
2022-05-12 12:52:29.2730 Info Validating config: TargetNames=logfile_mylog, ConfigItems=17
2022-05-12 12:52:29.2730 Debug Unused target checking is started... Rule Count: 1, Target Count: 1
2022-05-12 12:52:29.2730 Debug Unused target checking is completed. Total Rule Count: 1, Total Target Count: 1, Unused Target Count: 0
2022-05-12 12:52:29.2730 Debug Targets for MY.NAMESPACE.CLASS by level:
2022-05-12 12:52:29.2730 Debug Trace =>
2022-05-12 12:52:29.2730 Debug Debug => logfile_mylog
2022-05-12 12:52:29.2730 Debug Info => logfile_mylog
2022-05-12 12:52:29.2730 Debug Warn => logfile_mylog
2022-05-12 12:52:29.2730 Debug Error => logfile_mylog
2022-05-12 12:52:29.2730 Debug Fatal => logfile_mylog
2022-05-12 12:52:29.2730 Debug FileTarget(Name=logfile_mylog): Preparing for new file 'F:\LOGS\MYLOG.log'
2022-05-12 12:52:29.2730 Debug Creating file appender: F:\LOGS\MYLOG.log
2022-05-12 12:52:29.2730 Debug Setting 'NLog.LayoutRenderers.Wrappers.UppercaseLayoutRendererWrapper.uppercase' to 'true'
2022-05-12 12:52:29.2730 Debug Setting 'NLog.LayoutRenderers.DateLayoutRenderer.format' to 'yyyy-MM-dd HH:mm:ss'
2022-05-12 12:52:29.2730 Debug Setting 'NLog.LayoutRenderers.Wrappers.UppercaseLayoutRendererWrapper.uppercase' to 'true'
2022-05-12 12:52:29.2730 Debug Setting 'NLog.LayoutRenderers.Wrappers.PaddingLayoutRendererWrapper.padding' to '-5'
2022-05-12 12:52:29.2730 Debug Registered target logfile_mylog: NLog.Targets.FileTarget
2022-05-12 12:52:29.2730 Info Closing old configuration.
2022-05-12 12:52:29.2730 Debug LogFactory Flush with timeout=15 secs
2022-05-12 12:52:29.2730 Debug Flush completed
2022-05-12 12:52:29.2730 Debug Closing logging configuration...
2022-05-12 12:52:29.2730 Debug Closing target 'File Target[logfile_mylog]'.
2022-05-12 12:52:29.2730 Debug FileAppender Closing Invalidate - F:\LOGS\MYLOG.log
2022-05-12 12:52:29.2730 Debug Closed target 'File Target[logfile_mylog]'.
2022-05-12 12:52:29.2730 Debug Finished closing logging configuration.
2022-05-12 12:52:29.2730 Debug --- NLog configuration dump ---
2022-05-12 12:52:29.2730 Debug Targets:
2022-05-12 12:52:29.2730 Debug File Target[logfile_mylog]
2022-05-12 12:52:29.2730 Debug Rules:
2022-05-12 12:52:29.2730 Debug logNamePattern: (:All) levels: [ Debug Info Warn Error Fatal ] appendTo: [ logfile_mylog ]
2022-05-12 12:52:29.2730 Debug --- End of NLog configuration dump ---
2022-05-12 12:52:29.2730 Info Validating config: TargetNames=logfile_mylog, ConfigItems=17
2022-05-12 12:52:29.2730 Debug Unused target checking is started... Rule Count: 1, Target Count: 1
2022-05-12 12:52:29.2730 Debug Unused target checking is completed. Total Rule Count: 1, Total Target Count: 1, Unused Target Count: 0
2022-05-12 12:52:29.2730 Debug Targets for MY.NAMESPACE.CLASS by level:
2022-05-12 12:52:29.2730 Debug Trace =>
2022-05-12 12:52:29.2730 Debug Debug => logfile_mylog
2022-05-12 12:52:29.2730 Debug Info => logfile_mylog
2022-05-12 12:52:29.2730 Debug Warn => logfile_mylog
2022-05-12 12:52:29.2730 Debug Error => logfile_mylog
2022-05-12 12:52:29.2730 Debug Fatal => logfile_mylog
2022-05-12 12:52:29.2730 Debug FileTarget(Name=logfile_mylog): Preparing for new file 'F:\LOGS\MYLOG.log'
2022-05-12 12:52:29.2730 Debug Creating file appender: F:\LOGS\MYLOG.log
2022-05-12 12:52:29.3100 Debug Setting 'NLog.LayoutRenderers.Wrappers.UppercaseLayoutRendererWrapper.uppercase' to 'true'
2022-05-12 12:52:29.3100 Debug Setting 'NLog.LayoutRenderers.DateLayoutRenderer.format' to 'yyyy-MM-dd HH:mm:ss'
2022-05-12 12:52:29.3100 Debug Setting 'NLog.LayoutRenderers.Wrappers.UppercaseLayoutRendererWrapper.uppercase' to 'true'
2022-05-12 12:52:29.3100 Debug Setting 'NLog.LayoutRenderers.Wrappers.PaddingLayoutRendererWrapper.padding' to '-5'
2022-05-12 12:52:29.3100 Debug Registered target logfile_mylog: NLog.Targets.FileTarget
2022-05-12 12:52:29.3100 Info Closing old configuration.
2022-05-12 12:52:29.3100 Debug LogFactory Flush with timeout=15 secs
2022-05-12 12:52:29.3100 Debug Flush completed
2022-05-12 12:52:29.3100 Debug Closing logging configuration...
2022-05-12 12:52:29.3100 Debug Closing target 'File Target[logfile_mylog]'.
2022-05-12 12:52:29.3100 Debug FileAppender Closing Invalidate - F:\LOGS\MYLOG.log
2022-05-12 12:52:29.3100 Debug Closed target 'File Target[logfile_mylog]'.
2022-05-12 12:52:29.3100 Debug Finished closing logging configuration.
2022-05-12 12:52:29.3100 Debug --- NLog configuration dump ---
2022-05-12 12:52:29.3100 Debug Targets:
2022-05-12 12:52:29.3100 Debug File Target[logfile_mylog]
2022-05-12 12:52:29.3100 Debug Rules:
2022-05-12 12:52:29.3100 Debug logNamePattern: (:All) levels: [ Debug Info Warn Error Fatal ] appendTo: [ logfile_mylog ]
2022-05-12 12:52:29.3100 Debug --- End of NLog configuration dump ---
2022-05-12 12:52:29.3100 Info Validating config: TargetNames=logfile_mylog, ConfigItems=17
2022-05-12 12:52:29.3100 Debug Unused target checking is started... Rule Count: 1, Target Count: 1
2022-05-12 12:52:29.3100 Debug Unused target checking is completed. Total Rule Count: 1, Total Target Count: 1, Unused Target Count: 0
2022-05-12 12:52:29.3100 Debug Targets for MY.NAMESPACE.CLASS by level:
2022-05-12 12:52:29.3100 Debug Trace =>
2022-05-12 12:52:29.3100 Debug Debug => logfile_mylog
2022-05-12 12:52:29.3100 Debug Info => logfile_mylog
2022-05-12 12:52:29.3100 Debug Warn => logfile_mylog
2022-05-12 12:52:29.3100 Debug Error => logfile_mylog
2022-05-12 12:52:29.3100 Debug Fatal => logfile_mylog
2022-05-12 12:52:29.3100 Debug FileTarget(Name=logfile_mylog): Preparing for new file 'F:\LOGS\MYLOG.log'
2022-05-12 12:52:29.3100 Debug Creating file appender: F:\LOGS\MYLOG.log
2022-05-12 12:52:29.3200 Debug Setting 'NLog.LayoutRenderers.Wrappers.UppercaseLayoutRendererWrapper.uppercase' to 'true'
2022-05-12 12:52:29.3200 Debug Setting 'NLog.LayoutRenderers.DateLayoutRenderer.format' to 'yyyy-MM-dd HH:mm:ss'
2022-05-12 12:52:29.3200 Debug Setting 'NLog.LayoutRenderers.Wrappers.UppercaseLayoutRendererWrapper.uppercase' to 'true'
2022-05-12 12:52:29.3200 Debug Setting 'NLog.LayoutRenderers.Wrappers.PaddingLayoutRendererWrapper.padding' to '-5'
2022-05-12 12:52:29.3200 Debug Registered target logfile_mylog: NLog.Targets.FileTarget
2022-05-12 12:52:29.3200 Info Closing old configuration.
2022-05-12 12:52:29.3200 Debug LogFactory Flush with timeout=15 secs
2022-05-12 12:52:29.3200 Debug Flush completed
2022-05-12 12:52:29.3200 Debug Closing logging configuration...
2022-05-12 12:52:29.3200 Debug Closing target 'File Target[logfile_mylog]'.
2022-05-12 12:52:29.3200 Debug FileAppender Closing Invalidate - F:\LOGS\MYLOG.log
2022-05-12 12:52:29.3200 Debug Closed target 'File Target[logfile_mylog]'.
2022-05-12 12:52:29.3200 Debug Finished closing logging configuration.
2022-05-12 12:52:29.3200 Debug --- NLog configuration dump ---
2022-05-12 12:52:29.3200 Debug Targets:
2022-05-12 12:52:29.3200 Debug File Target[logfile_mylog]
2022-05-12 12:52:29.3200 Debug Rules:
2022-05-12 12:52:29.3200 Debug logNamePattern: (:All) levels: [ Debug Info Warn Error Fatal ] appendTo: [ logfile_mylog ]
2022-05-12 12:52:29.3200 Debug --- End of NLog configuration dump ---
2022-05-12 12:52:29.3200 Info Validating config: TargetNames=logfile_mylog, ConfigItems=17
2022-05-12 12:52:29.3200 Debug Unused target checking is started... Rule Count: 1, Target Count: 1
2022-05-12 12:52:29.3200 Debug Unused target checking is completed. Total Rule Count: 1, Total Target Count: 1, Unused Target Count: 0
2022-05-12 12:52:29.3200 Debug Targets for MY.NAMESPACE.CLASS by level:
2022-05-12 12:52:29.3200 Debug Trace =>
2022-05-12 12:52:29.3200 Debug Debug => logfile_mylog
2022-05-12 12:52:29.3200 Debug Info => logfile_mylog
2022-05-12 12:52:29.3200 Debug Warn => logfile_mylog
2022-05-12 12:52:29.3200 Debug Error => logfile_mylog
2022-05-12 12:52:29.3200 Debug Fatal => logfile_mylog
2022-05-12 12:52:29.3200 Debug FileTarget(Name=logfile_mylog): Preparing for new file 'F:\LOGS\MYLOG.log'
2022-05-12 12:52:29.3200 Debug Creating file appender: F:\LOGS\MYLOG.log

```

**The current result:** 
The logs is correctly written until a specific time, where the process seems to crash and keep the hand.
The log is in a workflow and the workflow is stuck after the crash happened.

**The expected result:**
No crash, and if crash, correctly return the error to the parent process

**Any current work-arounds:** 
None

**The current config (file content or API calls). :**
```
		private static readonly Logger _logger = LogManager.GetCurrentClassLogger();

        public CLASS()
        {
            var folderLogs = Path.Combine(@"F:\", "LOGS");
            var config = new LoggingConfiguration();
            var fileTarget = new FileTarget
            {
                Name = "logfile_mylog",
                Layout = "[${date:format=yyyy-MM-dd HH\\:mm\\:ss}] [${level:uppercase=true:padding=-5}] ${message}",
                FileName = Path.Combine(folderLogs, "MYLOG.log"),
                ArchiveEvery = FileArchivePeriod.Day,
                ArchiveFileName = Path.Combine(folderLogs, "MYLOG-{########}.log"),
                ArchiveNumbering = ArchiveNumberingMode.Date,
                ArchiveDateFormat = "yyyyMMdd",
                MaxArchiveFiles = 5
            };
            config.AddRule(LogLevel.Debug, LogLevel.Fatal, fileTarget, "*");
            LogManager.Configuration = config;

            // enable internal logging to a file
            NLog.Common.InternalLogger.LogFile = Path.Combine(folderLogs, "nlog-internal.txt"); 
            // set internal log level
            NLog.Common.InternalLogger.LogLevel = LogLevel.Debug;

            _logger.Info("Exit init");
        }
```

**Comment by snakefoot** [comment]
Surprised that an unresolved-type-event causes the entire application to halt. Have now created #4919, that will avoid type-resolve-event unless running on very old .NET Framework (Released with NLog 5.0).

But sadly enough then I guess it will not directly help for .NET 3,5 Workflow. But with NLog 5.0 then FileTarget will by default use ConcurrentWrites=false, which stops NLog from using operating-system-mutex for multi-process-services.

So my advice is to use NLog 5.0 RC2 until NLog 5.0 RTM is released. Or try to disable/clear the System.Diagnostics.TraceListeners that seems to cause havoc on `AppDomain.TypeResolveEvent`

**Comment by sylvainclb** [comment]
Hello snakefoot,

I dig more on my issue, and I confirm that the crash is not related to NLog, as it's happening without using it.
It was confusing as the the only error was ```Unable to resolve type: Mono.Runtime.```

Thanks for our help, I will look to use NLog 5.0 RC2
